rain_api_core/auth.py

```diff
@@ -0,0 +1,154 @@
+import contextlib
+import dataclasses
+import json
+import urllib.request
+from http.cookies import CookieError, SimpleCookie
+from time import time
+from typing import List, Mapping, Union
+from wsgiref.handlers import format_date_time as format_7231_date
+
+import jwt
+
+from rain_api_core.logging import get_log
+
+log = get_log()
+
+
+@dataclasses.dataclass
+class UserProfile:
+    user_id: str
+    token: str
+    groups: List[str]
+    first_name: str
+    last_name: str
+    email: str
+    iat: int = None
+    exp: int = None
+
+    @staticmethod
+    def from_jwt_payload(payload):
+        return UserProfile(
+            user_id=payload.get('urs-user-id'),
+            token=payload.get('urs-access-token'),
+            groups=payload.get('urs-groups'),
+            first_name=payload.get('first_name'),
+            last_name=payload.get('last_name'),
+            email=payload.get('email'),
+            iat=payload.get('iat'),
+            exp=payload.get('exp')
+        )
+
+
+class JwtManager:
+    def __init__(
+            self,
+            algorithm: str,
+            public_key: str,
+            private_key: str,
+            cookie_name: str,
+            blacklist_endpoint,
+            session_ttl_in_hours: int = 7 * 24
+    ):
+        self.algorithm = algorithm
+        self.public_key = public_key
+        self.private_key = private_key
+        self.cookie_name = cookie_name
+        self.black_list_endpoint = blacklist_endpoint
+        self.session_ttl = session_ttl_in_hours * 60 * 60
+        self.black_list_last_retrieved = 0
+        self.black_list = {}
+
+    def _get_auth_cookie(self, headers: Mapping[str, str]):
+        cookie_string = headers.get('cookie') or headers.get('Cookie') or headers.get('COOKIE')
+        if not cookie_string:
+            return {}
+
+        cookie = SimpleCookie()
+        with contextlib.suppress(CookieError):
+            cookie.load(cookie_string)
+        return cookie.get(self.cookie_name)
+
+    def _decode_jwt(self, token: str):
+        try:
+            return jwt.decode(token.encode(), self.public_key, self.algorithm)
```

Comment:
BTW, the definition of `jwt.decode` actually looks like this:

```python3
    def decode(
        jwt: str,
        key: str = "",
        algorithms: List[str] = None,
        options: Dict = None,
        **kwargs,
    ) -> Dict[str, Any]:
``` 

Notice that `algorithms` is a list of allowed algorithms. It seems that passing a single string value doesn't cause any errors, but it may be better to wrap it in a list to be future proof.